### ISSUE TITLE
Abstract ReadOnlyCollection

### ISSUE THREAD

**Issue by PawcioCNCProfi** [MicrosoftDocs/visualstudio-docs#4322]
I have a setup like this:
```
/// <summary>Base class to types which define an option for command line.</summary>
public abstract class CMDOption {
	/// <summary>Specifies list of available option names that can be used as command line arguments to recognize an option.</summary>
	public abstract ReadOnlyCollection<string> names { get; protected set; }
	//...
}
```

I don't get why it shows the warning under `names`.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8db6af07-b7d6-9395-223e-1a0f83fb2405
* Version Independent ID: 5ee87730-ff01-2571-494e-91d29c5fe211
* Content: [CA2227: Collection properties should be read only - Visual Studio](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2227?view=vs-2019)
* Content Source: [docs/code-quality/ca2227.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/code-quality/ca2227.md)
* Product: **visual-studio-windows**
* Technology: **vs-ide-code-analysis**
* GitHub Login: @jillre
* Microsoft Alias: **jillfra**

**Comment by WilliamAntonRohm** [comment]
@PawcioCNCProfi -- Thank you for your feedback.  Here are a few options where you might consider asking your question:​  

- [https://social.msdn.microsoft.com/Forums/vstudio](https://social.msdn.microsoft.com/Forums/vstudio) 
- [https://www.stackoverflow.com](https://www.stackoverflow.com) 
- [https://www.reddit.com/r/VisualStudio/](https://www.reddit.com/r/VisualStudio) 

As well, you might also consider using the [Report a Problem](https://docs.microsoft.com/visualstudio/ide/how-to-report-a-problem-with-visual-studio) tool to report the issue. 

@jillre -- please look into this potential documentation issue.

**Comment by Mikejo5000** [comment]
@PawcioCNCProfi What is the behavior you are expecting in this case? In the doc, for the fix, I don't see mention of using 'protected set', but just this:
"If replacing the collection is a goal, the preferred design pattern is to include a method to remove all the elements from the collection, and a method to repopulate the collection..."

**Comment by PawcioCNCProfi** [comment]
I expect the warning is not shown is such case. As it's described, the concern of this rule is that

> A writable collection property allows a user to replace the collection with a completely different collection.

 But the user can't change the ReadOnlyCollection<T> externally and the implementator have a full control over the collection anyway.

I could make the collection private and craft a method for the implementator to define collection only once, but this is still does not guaranteed that it will be used correctly and it makes more uncomfortable to implement the class...

**Comment by Mikejo5000** [comment]
@mavasani Should a ReadOnlyCollection generate CA2227 in this example?

**Comment by mavasani** [comment]
Can you port this issue to https://github.com/dotnet/roslyn-analyzers?

**Comment by Mikejo5000** [comment]
Closing in favor of the linked product issue.
